torch/nested/_internal/ops.py

```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+    Performs a sum along the provided tensor dimension.
+    Returns a dense tensor if the ragged dimension is reduced away, else returns a nested tensor.
+    """
     _, new_kwargs = normalize_function(
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
     inp = new_kwargs.pop("input")
-    assert inp._ragged_idx == 1
-    new_kwargs["dim"], ragged_reduced_away = _wrap_jagged_dims(
-        inp.dim(), new_kwargs["dim"], "sum"
+
+    (
+        new_kwargs["dim"],
+        reduce_on_batch,
+        reduce_on_ragged,
+        reduce_on_non_batch,  # noqa: UFMT
+    ) = _wrap_jagged_dims(
+        inp.dim(),
+        new_kwargs["dim"],
+        "sum",
+        inp._ragged_idx,
     )
 
-    if not ragged_reduced_away:
-        return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
-    else:
-        # Don't wrap because we reduced away the raggedness
-        out = func(inp._values, **new_kwargs)
+    if inp._ragged_idx != 1:
+        raise RuntimeError("sum(): not supported when ragged_idx != 1 for NestedTensor")
+
+    if reduce_on_ragged and inp._lengths is not None:
+        raise RuntimeError(
+            "sum(): not supported for nested tensors with holes, i.e. when lengths is not None, "
+            + "if reducing across the ragged dimension for NestedTensor"
+        )
+
+    if reduce_on_ragged:  # raggedness reduced away --> return dense tensor
+        if (
+            reduce_on_batch
+        ):  # reduction cases: (batch, ragged), (batch, ragged, non-batch), etc.
+            out = func(
+                inp._values, **new_kwargs
+            )  # no need to read offsets --> apply sum directly on values
+        else:
+            if (
+                reduce_on_non_batch
+            ):  # invalid reduction cases: (ragged, non-batch), etc.
+                raise RuntimeError(
+                    "sum(): not supported along a ragged and non-batch dimension for NestedTensor"
+                )
+            # reduction cases: (ragged)
+            out = torch.sum(
+                torch.ops.aten._jagged_to_padded_dense_forward(
+                    inp._values.view(*inp._values.shape[: inp._ragged_idx], -1),
+                    [inp._offsets],
+                    max_lengths=[inp._max_seqlen],
+                ).view(
+                    *inp.shape[: inp._ragged_idx],
+                    inp._max_seqlen,
+                    *inp.shape[inp._ragged_idx + 1 :],
+                ),
+                dim=inp._ragged_idx,
+            )  # need to read offsets --> pad jagged dimension and apply sum
+
         if new_kwargs["keepdim"]:
             out = out.unsqueeze(0)
         return out
+    else:  # raggedness preserved --> return nested tensor
+        if (
+            reduce_on_batch
+        ):  # invalid reduction cases: (batch), (batch, non-batch), etc.
+            raise RuntimeError(
+                "sum(): not supported along a batch dimension for NestedTensor"
```

Comment:
nit: maybe also mention that reduction is supported when batch dim + ragged dim are reduced

Reply:
i do mention this in a [comment](https://github.com/pytorch/pytorch/pull/130425/files#diff-5ab78b751bd84fe009dbdb8899eb434a0278315bc33f15dfe45d8d87a2afa662R883) in my code; however, would it be clearer to modify this to `sum(): not supported along only the batch dimension for NestedTensor`?